.github/workflows/generate_sbom.py

```diff
@@ -0,0 +1,20 @@
+from json import dumps
+
+from requests import get
+
+with open("opentelemetry-python.spdx.json", "w") as sbom_file:
+    sbom_file.write(
+        dumps(
+            get(
+                (
+                    "https://api.github.com/repos/open-telemetry/"
+                    "opentelemetry-python/dependency-graph/sbom"
```

Comment:
This API seems pretty barebones, I'm guessing this just generates an SBOM against the main branch and maybe cached at some random commit?

- What if you are making a patch release, would the dep graph be correct?
- If it is looking at main branch, I think our release process would show the `.dev0` dependencies instead of actual released versions

Reply:
> This API seems pretty barebones, I'm guessing this just generates an SBOM against the main branch and maybe cached at some random commit?

No, apparently this functionality creates an SBOM file using the latest commit in `main`. There is a way to generate this file using the web interface (`https://github.com/open-telemetry/opentelemetry-python`/three dots button in the upper right corner/Insights/Dependency graph/Export SBOM). This creates a file that includes the latest commit from `main` in its name. I assume the same file gets generated when the REST API is being used.

Reply:
> What if you are making a patch release, would the dep graph be correct? If it is looking at main branch, I think our release process would show the .dev0 dependencies instead of actual released versions

The reason why the SBOM file includes released versions is because we have `requirements.txt`-like files that include released versions. For example, I made these changes, and pushed them to `main`:

```diff
diff --git a/docs/examples/metrics/reader/requirements.txt b/docs/examples/metrics/reader/requirements.txt
index 2ccffaf3..23ff5bf6 100644
--- a/docs/examples/metrics/reader/requirements.txt
+++ b/docs/examples/metrics/reader/requirements.txt
@@ -1,5 +1,5 @@
 Deprecated==1.2.13
-opentelemetry-api==1.15.0
+opentelemetry-api==1.20.0
 opentelemetry-sdk==1.15.0
 opentelemetry-semantic-conventions==0.36b0
 typing_extensions==4.3.0
diff --git a/docs/examples/metrics/views/requirements.txt b/docs/examples/metrics/views/requirements.txt
index be612711..d9590b3f 100644
--- a/docs/examples/metrics/views/requirements.txt
+++ b/docs/examples/metrics/views/requirements.txt
@@ -1,5 +1,5 @@
 Deprecated==1.2.13
-opentelemetry-api==1.12.0
+opentelemetry-api==1.20.0
 opentelemetry-sdk==1.12.0
 opentelemetry-semantic-conventions==0.33b0
 typing_extensions==4.3.0
```

Those changes produced the following changes in the SBOM file:

```diff
5c5
<     "created": "2024-04-03T19:17:07Z",
---
>     "created": "2024-04-03T17:15:55Z",
16c16
<   "documentNamespace": "https://github.com/SecuringCarter/opentelemetry-python/dependency_graph/sbom-7c05aa1b812baa89",
---
>   "documentNamespace": "https://github.com/SecuringCarter/opentelemetry-python/dependency_graph/sbom-75a76526c2dd97b4",
814a815,830
>       "SPDXID": "SPDXRef-pip-opentelemetry-api-1.15.0",
>       "name": "pip:opentelemetry-api",
>       "versionInfo": "1.15.0",
>       "downloadLocation": "NOASSERTION",
>       "filesAnalyzed": false,
>       "licenseConcluded": "Apache-2.0",
>       "supplier": "NOASSERTION",
>       "externalRefs": [
>         {
>           "referenceCategory": "PACKAGE-MANAGER",
>           "referenceLocator": "pkg:pypi/opentelemetry-api@1.15.0",
>           "referenceType": "purl"
>         }
>       ]
>     },
>     {
878a895,910
>       "SPDXID": "SPDXRef-pip-opentelemetry-api-1.12.0",
>       "name": "pip:opentelemetry-api",
>       "versionInfo": "1.12.0",
>       "downloadLocation": "NOASSERTION",
>       "filesAnalyzed": false,
>       "licenseConcluded": "Apache-2.0",
>       "supplier": "NOASSERTION",
>       "externalRefs": [
>         {
>           "referenceCategory": "PACKAGE-MANAGER",
>           "referenceLocator": "pkg:pypi/opentelemetry-api@1.12.0",
>           "referenceType": "purl"
>         }
>       ]
>     },
>     {
2317a2350,2354
>       "relatedSpdxElement": "SPDXRef-pip-opentelemetry-api-1.15.0"
>     },
>     {
>       "relationshipType": "DEPENDS_ON",
>       "spdxElementId": "SPDXRef-com.github.SecuringCarter-opentelemetry-python",
2333a2371,2375
>     },
>     {
>       "relationshipType": "DEPENDS_ON",
>       "spdxElementId": "SPDXRef-com.github.SecuringCarter-opentelemetry-python",
>       "relatedSpdxElement": "SPDXRef-pip-opentelemetry-api-1.12.0"
```

So, I now think that we probably should do something about these `requirements-txt` files that do not actually represent project dependencies.

Reply:
After [many more changes](https://github.com/open-telemetry/opentelemetry-python/commit/877fb973b7d8a6f8f74733613b33b22d47df7ed3) I managed to reduce the amount of packages in the SBOM file with an empty value of `versionInfo` to only one:

```
{
    "SPDXID": "SPDXRef-com.github.SecuringCarter-opentelemetry-python",
    "name": "com.github.SecuringCarter/opentelemetry-python",
    "versionInfo": "",
    "downloadLocation": "git+https://github.com/SecuringCarter/opentelemetry-python",
    "licenseDeclared": "Apache-2.0",
    "filesAnalyzed": false,
    "supplier": "NOASSERTION",
    "externalRefs": [
        {
            "referenceCategory": "PACKAGE-MANAGER",
            "referenceType": "purl",
            "referenceLocator": "pkg:github/SecuringCarter/opentelemetry-python"
        }
    ]
}
```

I think the mechanism to generate an SBOM file that this PR introduces is valid. We probably need to change our repo in order to produce a "better" or more useful SBOM file.

Can we merge this and work on making the SBOM file better in subsequent PRs?